tooling/noir_js/test/backend/barretenberg.ts

```diff
@@ -0,0 +1,122 @@
+// @ts-ignore
+import { Barretenberg, Crs, RawBuffer } from '@aztec/bb.js';
+// TODO: This should be re-exported from @aztec/bb-js
+import { Ptr } from '@aztec/bb.js/dest/browser/types';
+import { acirToUint8Array } from '../../src/index.js';
+
+export class Backend {
+  // These type assertions are used so that we don't
+  // have to initialize `api` and `acirComposer` in the constructor.
+  // These are initialized asynchronously in the `init` function,
+  // constructors cannot be asynchronous which is why we do this.
+  api = {} as Barretenberg;
+  acirComposer = {} as Ptr;
+  acirUncompressedBytecode: Uint8Array;
+
+  constructor(acirBytecodeBase64: string) {
+    this.acirUncompressedBytecode = acirToUint8Array(acirBytecodeBase64);
+  }
+  async init() {
+    const numThreads = 4;
+
+    const { api, composer } = await this.initBarretenberg(numThreads, this.acirUncompressedBytecode);
+
+    this.api = api;
+    this.acirComposer = composer;
+  }
+
+  async initBarretenberg(numThreads: number, acirUncompressedBytecode: Uint8Array) {
+    const api = await Barretenberg.new(numThreads);
+
+    const [_exact, _total, subgroupSize] = await api.acirGetCircuitSizes(acirUncompressedBytecode);
+    const crs = await Crs.new(subgroupSize + 1);
+    await api.commonInitSlabAllocator(subgroupSize);
+    await api.srsInitSrs(new RawBuffer(crs.getG1Data()), crs.numPoints, new RawBuffer(crs.getG2Data()));
+
+    const acirComposer = await api.acirNewAcirComposer(subgroupSize);
+    return { api: api, composer: acirComposer };
+  }
+
+  // Generate an outer proof. This is the proof for the circuit which will verify
+  // inner proofs and or can be seen as the proof created for regular circuits.
+  //
+  // The settings for this proof are the same as the settings for a "normal" proof
+  // ie one that is not in the recursive setting.
+  async generateOuterProof(decompressedWitness: Uint8Array) {
+    const makeEasyToVerifyInCircuit = false;
+    return this.generateProof(decompressedWitness, makeEasyToVerifyInCircuit);
+  }
+
+  // Generates an inner proof. This is the proof that will be verified
+  // in another circuit.
+  //
+  // This is sometimes referred to as a recursive proof.
+  // We avoid this terminology as the only property of this proof
+  // that matters, is the fact that it is easy to verify in another
+  // circuit. We _could_ choose to verify this proof in the CLI.
+  //
+  // We set `makeEasyToVerifyInCircuit` to true, which will tell the backend to
+  // generate the proof using components that will make the proof
+  // easier to verify in a circuit.
+  async generateInnerProof(witness: Uint8Array) {
+    const makeEasyToVerifyInCircuit = true;
+    return this.generateProof(witness, makeEasyToVerifyInCircuit);
+  }
+
+  async generateProof(decompressedWitness: Uint8Array, makeEasyToVerifyInCircuit: boolean) {
```

Comment:
`makeEasyToVerifyInCircuit` is a bit weird of a name to me. I know what it means but maybe we just label it `inner_proof`? We can bikeshed this more later and is non-blocking